lib/p2p/Pool.ts

```diff
@@ -501,11 +507,44 @@ class Pool extends EventEmitter {
     }
 
     const peerPubKey = peer.nodePubKey!;
-    this.logger.verbose(`opened connection to ${peerPubKey} at ${addressUtils.toString(peer.address)}`);
-    this.pendingOutboundPeers.delete(peerPubKey);
+
+    // A potential race condition exists here in the case where two peers attempt connections
+    // to each other simultaneously. Handshakes may be completed on both sockets without being
+    // detected in the validatePeer method that is called above after a handshake begins and
+    // the node pub key for the inbound peer is established but before it completes. In this case,
+    // both peers cannot close the "duplicate" connection or else they may wind up closing both
+    // of them. The approach below arbitrarily picks the peer with the higher node pubkey to close
+    // the socket when it encounters a duplicate, while the other peer waits briefly to see if
+    // the already established connection is closed.
+    if (this.peers.has(peerPubKey)) {
+      if (this.nodePubKey > peerPubKey) {
+        await peer.close(DisconnectionReason.AlreadyConnected);
+        throw errors.NODE_ALREADY_CONNECTED(peerPubKey, peer.address);
+      } else {
+        const stillConnected = await new Promise<boolean>((resolve) => {
+          this.peers.get(peerPubKey)!.once('close', () => {
+            resolve(false);
+            clearTimeout(timeout);
+          });
+          const timeout = setTimeout(() => resolve(true), Peer.STALL_INTERVAL);
```

Comment:
Can we move the timeout declaration above? Reading from top down I got confused because `clearTimeout(timeout)`, but didn't see it declared.

Reply:
No problem.